airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/data/AirbyteValue.kt

```diff
@@ -5,27 +5,154 @@
 
-sealed interface AirbyteValue
+sealed interface AirbyteValue {
+    companion object {
+        fun from(value: Any?): AirbyteValue =
+            when (value) {
+                null -> NullValue
+                is String -> StringValue(value)
+                is Boolean -> BooleanValue(value)
+                is Int -> IntegerValue(value.toLong())
+                is Long -> IntegerValue(value)
+                is Double -> NumberValue(BigDecimal.valueOf(value))
+                is BigDecimal -> NumberValue(value)
+                is LocalDate -> DateValue(value.toString())
+                is OffsetDateTime,
+                is LocalDateTime -> TimestampValue(value.toString())
+                is OffsetTime,
+                is LocalTime -> TimeValue(value.toString())
+                is Map<*, *> ->
+                    ObjectValue.from(@Suppress("UNCHECKED_CAST") (value as Map<String, Any?>))
+                is List<*> -> ArrayValue.from(value)
+                else ->
+                    throw IllegalArgumentException(
+                        "Unrecognized value (${value.javaClass.name}: $value"
+                    )
+            }
+    }
+}
 
-data object NullValue : AirbyteValue
+// Comparable implementations are intended for use in tests.
+// They're not particularly robust, and probably shouldn't be relied on
+// for actual sync-time logic.
+// (mostly the date/timestamp/time types - everything else is fine)
+data object NullValue : AirbyteValue, Comparable<NullValue> {
+    override fun compareTo(other: NullValue): Int = 0
+}
 
-@JvmInline value class StringValue(val value: String) : AirbyteValue
+@JvmInline
+value class StringValue(val value: String) : AirbyteValue, Comparable<StringValue> {
+    override fun compareTo(other: StringValue): Int = value.compareTo(other.value)
+}
 
-@JvmInline value class BooleanValue(val value: Boolean) : AirbyteValue
+@JvmInline
+value class BooleanValue(val value: Boolean) : AirbyteValue, Comparable<BooleanValue> {
+    override fun compareTo(other: BooleanValue): Int = value.compareTo(other.value)
+}
 
-@JvmInline value class IntegerValue(val value: Long) : AirbyteValue
+@JvmInline
+value class IntegerValue(val value: Long) : AirbyteValue, Comparable<IntegerValue> {
+    override fun compareTo(other: IntegerValue): Int = value.compareTo(other.value)
+}
 
-@JvmInline value class NumberValue(val value: BigDecimal) : AirbyteValue
+@JvmInline
+value class NumberValue(val value: BigDecimal) : AirbyteValue, Comparable<NumberValue> {
+    override fun compareTo(other: NumberValue): Int = value.compareTo(other.value)
+}
 
-@JvmInline value class DateValue(val value: String) : AirbyteValue
+@JvmInline
+value class DateValue(val value: String) : AirbyteValue, Comparable<DateValue> {
+    override fun compareTo(other: DateValue): Int {
+        val thisDate =
+            try {
+                LocalDate.parse(value)
+            } catch (e: Exception) {
+                LocalDate.MIN
```

Comment:
The point of this is just to group the PKs for testing right? Any reason we can't just use the default comparators?

Reply:
grouping pk/cursor, yeah. String sort for temporal types is kind of iffy (`01:00-7` sorts lower than `07:00Z`, even though it's temporally later)

I could pull these out to an explicit `Comparator<AirbyteValue>` class instead of having AirbyteValue directly implement comparable? Which would make it easier to avoid people accidentally relying on these

Reply:
We can group that under the "abstract conversions from types work." That can include moving converters/comparators that are strictly for testing into the test area

(I'm tracking these btw here: https://docs.google.com/spreadsheets/d/1NXo7_A8ZyFkP3RKepGOhuNPL1i5o9ZqBFXFxSjF0ehE/edit?gid=0#gid=0; I'll work with @stephane-airbyte to prioritize w/r/t CDK project and w/ @jitenoswal to map them to work epics / ticket out etc)

Reply:
(or obviously if there's a convenient time to do it when doing test PRs)